lib/contextual_logger/logger_with_context.rb

```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+module ContextualLogger
+  # A logger that deep_merges additional context and then delegates to the given logger.
+  # Keeps it own log level that may be set independently of the logger it delegates to (the latter's log level is ignored).
+  class LoggerWithContext
+    include LoggerMixin
+
+    attr_accessor :level
+    attr_reader :context
+
+    def initialize(logger, context, level: nil)
+      @logger = logger
+      @level = level || logger.level
+      @context = context
+      @merged_context_cache = {}  # so we don't have to merge every time
+    end
+
+    def write_entry_to_log(severity, timestamp, progname, message, context)
+      if @merged_context_cache.size >= 5000 # keep this cache memory use finite
```

Comment:
this optimization can cause some pretty large memory bloat.  is it worth it?  is the same context being passed inline with the `info`, `debug`, etc. methods that we need to do this?  i think if that were the case, the best practice would be pushed forward that wrapping all those methods with what is similar being set using a `with_context` block so that what's not similar is the only pieces being passed inline.  by doing this best practice, there would be no cache hits on this.

Reply:
> this optimization can cause some pretty large memory bloat

Huh, the 5000 test is designed to not bloat memory. We could change that to 1000 or 500 or anything else to lessen the temporary bloat, if that's your concern.

> is it worth it? 

I think so. I suppose I could run a benchmark to prove/disprove it?

Reply:
I'm a little concerned about the memory bloat that this can cause, but I'd feel comfortable with it as long as we're running some reasonable benchmarks to make sure that the bloat isn't too large.

Reply:
I made two changes: 1) the limit is now 1000 instead of 5000 and 2) once we hit the limit, we just leave the cache as it was (stop growing it but also don't clear it). So our bloat is limited to 1000 entries and after that, it won't change. My reasoning is that the first 1000 are likely have included some good cache hit ratio values.
https://github.com/Invoca/contextual_logger/pull/12/commits/dad998249f5390c300bf6757a44e02497eee1e5e